google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+                bundled_document.metadata.queries.append(query_name)  # type: ignore
+        else:
+            raise ValueError(f"Unexpected type of BundleElement: {type}")
+
+    def build(self) -> str:
+        """Iterates over the bundle's stored documents and queries and produces
+        a single length-prefixed json string suitable for long-term storage.
+
+        Example:
+
+            from google.cloud import firestore
+
+            db = firestore.Client()
+            collection_ref = db.collection(u'users')
+
+            bundle = firestore.FirestoreBundle('my bundle')
+            bundle.add('app-users', collection_ref._query())
+
+            serialized_bundle: str = bundle.build()
+
+            # Now upload `serialized_bundle` to Google Cloud Storage, store it
+            # in Memorystore, or any other storage solution.
+
+        Returns:
+            str: The length-prefixed string representation of this bundle'
+                contents.
+        """
+        buffer: str = ""
+
+        named_query: NamedQuery
+        for named_query in self.named_queries.values():
+            buffer += self._compile_bundle_element(
+                BundleElement(named_query=named_query)
+            )
+
+        bundled_document: "_BundledDocument"  # type: ignore
+        document_count: int = 0
+        for bundled_document in self.documents.values():
+            buffer += self._compile_bundle_element(
+                BundleElement(document_metadata=bundled_document.metadata)
+            )
+            document_count += 1
+            buffer += self._compile_bundle_element(
+                BundleElement(document=bundled_document.snapshot._to_protobuf()._pb,)
+            )
+
+        metadata: BundleElement = BundleElement(
+            metadata=self._deserialized_metadata
+            or BundleMetadata(
+                id=self.name,
+                create_time=_helpers.build_timestamp(),
+                version=FirestoreBundle.BUNDLE_SCHEMA_VERSION,
+                total_documents=document_count,
+                total_bytes=len(buffer.encode("utf-8")),
+            )
+        )
+        return f"{self._compile_bundle_element(metadata)}{buffer}"
+
+    def _compile_bundle_element(self, bundle_element: BundleElement) -> str:
+        serialized_be: str = json.dumps(BundleElement.to_dict(bundle_element))
+        return f"{len(serialized_be)}{serialized_be}"
+
+    def get_documents(
+        self, *, query_name: Optional[str] = None
+    ) -> Generator[DocumentSnapshot, None, None]:
+        bundled_doc: "_BundledDocument"
+        for bundled_doc in self.documents.values():
+            if query_name and query_name not in bundled_doc.metadata.queries:
+                continue
+            yield bundled_doc.snapshot
+
+    def get_document(self, document_id: str) -> Optional[DocumentSnapshot]:
+        bundled_doc = self.documents.get(document_id)
+        if bundled_doc:
+            return bundled_doc.snapshot
```

Comment:
I am getting somewhat hesitant about these methods. I don't see a  clear use case for them (other than our unit tests) and they don't exist on other Server or our Mobile SDKs, which creates API discrepancies and more work in the long run because one or two user will eventually ask us to add these APIs everywhere.

Reply:
+1

Reply:
What is the intended way that developers would access the content within a Bundle? It feels like developers will either need to exist or will write this same code on their own.

Am I missing anything there?

Reply:
Moved these functions to `_helpers`.